src/main/java/gyro/google/compute/AbstractDiskResource.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    /**
+     * The fully-qualified URL linking back to the disk.
+     */
+    @Output
+    public String getSelfLink() {
+        return selfLink;
+    }
+
+    public void setSelfLink(String selfLink) {
+        this.selfLink = selfLink;
+    }
+
+    /**
+     * Links to the attached instances of the disk.
+     */
+    @Output
+    public List<String> getUsers() {
+        if (users == null) {
+            users = new ArrayList<>();
+        }
+        return users;
+    }
+
+    public void setUsers(List<String> users) {
+        this.users = users;
+    }
+
+    /**
+     * The fingerprint for the labels being applied to the disk, which is updated any time the labels change.
+     */
+    @Output
+    public String getLabelFingerprint() {
+        return labelFingerprint;
+    }
+
+    public void setLabelFingerprint(String labelFingerprint) {
+        this.labelFingerprint = labelFingerprint;
+    }
+
+    @Override
+    public void copyFrom(Disk disk) {
+        setName(disk.getName());
+        setDescription(disk.getDescription());
+        setSizeGb(disk.getSizeGb());
+        setSourceSnapshot(disk.getSourceSnapshot());
+        setLabels(disk.getLabels());
+        setPhysicalBlockSizeBytes(disk.getPhysicalBlockSizeBytes());
+        setStatus(disk.getStatus());
+        setSourceSnapshotId(disk.getSourceSnapshotId());
+        setSelfLink(disk.getSelfLink());
+        setUsers(disk.getUsers());
+        setLabelFingerprint(disk.getLabelFingerprint());
+
+        if (disk.getDiskEncryptionKey() != null) {
+            EncryptionKey diskEncryption = newSubresource(EncryptionKey.class);
+            diskEncryption.copyFrom(disk.getDiskEncryptionKey());
+            setDiskEncryptionKey(diskEncryption);
+        }
+
+        if (disk.getSourceSnapshotEncryptionKey() != null) {
+            EncryptionKey sourceSnapshotEncryption = newSubresource(EncryptionKey.class);
+            sourceSnapshotEncryption.copyFrom(disk.getSourceSnapshotEncryptionKey());
+            setSourceSnapshotEncryptionKey(sourceSnapshotEncryption);
+        }
+
+        copyMore(disk);
+    }
+
+    public abstract void copyMore(Disk disk);
+
+    @Override
+    public void create(GyroUI ui, State state) {
+        Disk disk = new Disk();
+        disk.setName(getName());
+        disk.setDescription(getDescription());
+        disk.setSizeGb(getSizeGb());
+        disk.setType(getType());
+        disk.setPhysicalBlockSizeBytes(getPhysicalBlockSizeBytes());
+        disk.setLabels(getLabels());
+        disk.setResourcePolicies(getResourcePolicies());
+        disk.setSourceSnapshot(getSourceSnapshot());
+        disk.setDiskEncryptionKey(getDiskEncryptionKey() != null ? getDiskEncryptionKey().toCustomerEncryptionKey() : null);
+        disk.setSourceSnapshotEncryptionKey(getSourceSnapshotEncryptionKey() != null
+            ? getSourceSnapshotEncryptionKey().toCustomerEncryptionKey()
+            : null);
+
+        doCreate(ui, state, disk);
+    }
```

Comment:
This is not actually creating anything and can be replaced by a `protected Disk toDisk() {}` and corresponding concrete class can call this method to get the disk object and use as required. This will also remove the need for the method `doCreate()`